### PR TITLE
stm32 updates

### PR DESCRIPTION
**Pull request by buha**
some stm32 related updates

none of these changes break anything already on master

### CODE REVIEW

**Comment by buha** [comment]
v2:
- cppcheck fix attempt

**Comment by buha** [comment]
v3:
- added an extra commit adding microsecond precision udelay using DWT (if available)
- fixed warnings coming from stm32_uart_stdio

**Comment by buha** [comment]
V4:
- replaced specific include stm32f4xx_hal.h with stm32_hal.h, leftover by mistake

File: drivers/platform/stm32/stm32_i2c.c [hunk 1]
```diff
@@ -70,9 +70,29 @@ int32_t stm32_i2c_init(struct i2c_desc **desc,
...
+	case 3:
+		base = I2C3;
+		break;
+#endif
+	default:
+		ret = -EINVAL;
+		goto error_2;
+	};
+
+	xdesc->hi2c.Instance = base;
```
Comment (adrimbarean) [inline drivers/platform/stm32/stm32_i2c.c:95]: nitpick: I'd say there is no need for the intermediate pointer 'base'. We can just assign 'xdesc->hi2c.Instance' in switch.

**Review by adrimbarean — APPROVED** [review]
Looks good to me. Just one comment.

**Review by amiclaus — APPROVED** [review]

**Pull request merged**